File: ios/loveaceios/Storage/ScheduleStore.swift
import Foundation

final class ScheduleStore: Sendable {
    private let defaults = UserDefaults.standard
    private static let prefsName = "schedule_cache"

    nonisolated var activeUserId: String {
        get { UserDefaults.standard.string(forKey: "schedule_active_user") ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: "schedule_active_user") }
    }

    private func key(_ base: String) -> String {
        let uid = activeUserId
        return uid.isEmpty ? base : "\(uid)_\(base)"
    }

    func saveCourses(_ courses: [ScheduleCourse]) {
        if let data = try? JSONEncoder().encode(courses) {
            defaults.set(data, forKey: key("courses_json"))
            defaults.set(Date().timeIntervalSince1970, forKey: key("updated_at"))
        }
    }

    func loadCourses() -> [ScheduleCourse] {
        guard let data = defaults.data(forKey: key("courses_json")) else { return [] }
        return (try? JSONDecoder().decode([ScheduleCourse].self, from: data)) ?? []
    }

    func hasCourses() -> Bool { defaults.data(forKey: key("courses_json")) != nil }

    func saveSemesterJSON(_ rawJSON: String) {
        defaults.set(rawJSON, forKey: key("semester_json"))
        defaults.set(Date().timeIntervalSince1970, forKey: key("semester_updated_at"))
    }

    func loadSemesterJSON() -> String? { defaults.string(forKey: key("semester_json")) }

    func clear() {
        defaults.removeObject(forKey: key("courses_json"))
        defaults.removeObject(forKey: key("updated_at"))
        defaults.removeObject(forKey: key("semester_json"))
        defaults.removeObject(forKey: key("semester_updated_at"))
    }
}

File: android/app/src/main/java/tech/loveace/appv3/data/local/ScheduleStore.kt
package tech.loveace.appv3.data.local

import android.content.Context
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import tech.loveace.appv3.data.model.ScheduleCourse

/**
 * 课程表 & 学期信息缓存 — SharedPreferences
 * 供小组件读取，不需要登录和网络
 * 按用户学号隔离存储
 */
class ScheduleStore(context: Context) {

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true }

    /** 当前活跃用户 ID（学号），由外部设置 */
    var activeUserId: String = ""

    private fun key(base: String) = if (activeUserId.isNotEmpty()) "${activeUserId}_$base" else base

    // ── 课程缓存 ──

    fun saveCourses(courses: List<ScheduleCourse>) {
        prefs.edit()
            .putString(key(KEY_COURSES), json.encodeToString(courses))
            .putLong(key(KEY_UPDATED_AT), System.currentTimeMillis())
            .apply()
    }

    fun loadCourses(): List<ScheduleCourse> {
        val data = prefs.getString(key(KEY_COURSES), null) ?: return emptyList()
        return try { json.decodeFromString<List<ScheduleCourse>>(data) } catch (_: Exception) { emptyList() }
    }

    fun hasCourses(): Boolean = prefs.contains(key(KEY_COURSES))

    // ── 学期信息缓存 ──

    fun saveSemesterJson(rawJson: String) {
        prefs.edit()
            .putString(key(KEY_SEMESTER), rawJson)
            .putLong(key(KEY_SEMESTER_UPDATED), System.currentTimeMillis())
            .apply()
    }

    fun loadSemesterJson(): String? = prefs.getString(key(KEY_SEMESTER), null)

    fun clear() {
        // 只清除当前用户的数据
        prefs.edit()
            .remove(key(KEY_COURSES))
            .remove(key(KEY_UPDATED_AT))
            .remove(key(KEY_SEMESTER))
            .remove(key(KEY_SEMESTER_UPDATED))
            .apply()
    }

    companion object {
        private const val PREFS_NAME = "schedule_cache"
        private const val KEY_COURSES = "courses_json"
        private const val KEY_UPDATED_AT = "updated_at"
        private const val KEY_SEMESTER = "semester_json"
        private const val KEY_SEMESTER_UPDATED = "semester_updated_at"
    }
}
